drv/stm32h7-sprot-server/src/main.rs

```diff
@@ -275,164 +288,147 @@ impl ServerImpl {
+        self.spi.read(buf).map_err(|_| SprotError::SpiServerError)?;
+
+        self.rx_buf.validate_crc(&header)?;
+
+        Ok(VerifiedRxMsg(header))
+    }
+
+    fn log_parse_error(&self) {
+        ringbuf_entry!(Trace::RxParseError(
+            self.rx_buf.as_slice()[0],
+            self.rx_buf.as_slice()[1],
+            self.rx_buf.as_slice()[2],
+            self.rx_buf.as_slice()[3]
+        ));
     }
 
     fn do_send_recv_retries(
         &mut self,
-        size: usize,
+        txmsg: VerifiedTxMsg,
         timeout: u32,
         retries: u16,
-    ) -> Result<(MsgType, usize), MsgError> {
+    ) -> Result<VerifiedRxMsg, SprotError> {
         let mut attempts_left = retries;
-        let mut errcode = MsgError::Unknown;
+        let mut errcode = SprotError::Unknown;
         loop {
             if attempts_left == 0 {
                 ringbuf_entry!(Trace::FailedRetries { retries, errcode });
                 break;
             }
             attempts_left -= 1;
 
-            match self.do_send_recv(size, timeout) {
+            match self.do_send_recv(txmsg, timeout) {
                 // Recoverable errors dealing with our ability to receive
                 // the message from the RoT.
-                Err(err) if err == MsgError::InvalidCrc => {
-                    errcode = err;
-                    continue;
-                }
-                Err(err)
-                    if matches!(
-                        err,
-                        MsgError::EmptyMessage
-                            | MsgError::RotNotReady
-                            | MsgError::RotBusy
-                    ) =>
-                {
-                    errcode = err;
-                    continue;
-                }
-                // The remaining errors are not recoverable.
                 Err(err) => {
-                    ringbuf_entry!(Trace::MsgError(err));
-                    errcode = err;
-                    break;
+                    ringbuf_entry!(Trace::SprotError(err));
+                    if is_recoverable_error(err) {
+                        errcode = err;
+                        continue;
+                    } else {
+                        return Err(err);
+                    }
                 }
+
                 // Intact messages from the RoT may indicate an error on
                 // its side.
-                Ok((msgtype, payload_len)) => {
-                    match msgtype {
-                        MsgType::ErrorRsp if payload_len > 0 => {
-                            let payload = payload_buf(
-                                Some(payload_len),
-                                &self.rx_buf[..],
-                            );
-                            errcode =
-                                MsgError::from_u8(payload[0]).unwrap_lite();
-                            ringbuf_entry!(Trace::MsgError(errcode));
+                Ok(rxmsg @ VerifiedRxMsg(header)) => {
```

Comment:
cursed syntax, but okay

Reply:
Actually, is there a reason to explicitly unpack the header, versus using `rxmsg.header` in the code below?

Reply:
It's confusing. I'll do what you suggest.

Reply:
Fixed in [3a0f61a](https://github.com/oxidecomputer/hubris/pull/965/commits/3a0f61ac0031aaed46524561bf56eb41f20f085f)